pkg/auth/providers/genericoidc/genericoidc_provider.go

```diff
@@ -0,0 +1,161 @@
+	return &GenOIDCProvider{
+		baseoidc.OpenIDCProvider{
+			Name:        Name,
+			Type:        client.GenericOIDCConfigType,
+			CTX:         ctx,
+			AuthConfigs: mgmtCtx.Management.AuthConfigs(""),
+			Secrets:     mgmtCtx.Core.Secrets(""),
+			UserMGR:     userMGR,
+			TokenMGR:    tokenMGR,
+		},
+	}
+}
+
+// GetName returns the name of this provider.
+func (g *GenOIDCProvider) GetName() string {
+	return Name
+}
+
+// SearchPrincipals will return a principal of the requested principalType with a displayName and loginName
+// that match the searchValue.  This is done because OIDC does not have a proper lookup mechanism.  In order
+// to provide some degree of functionality that allows manual entry for users/groups, this is the compromise.
+func (g *GenOIDCProvider) SearchPrincipals(searchValue, principalType string, _ v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+
+	if principalType == "" {
+		principalType = UserType
+	}
+
+	p := v3.Principal{
+		ObjectMeta:    metav1.ObjectMeta{Name: g.Name + "_" + principalType + "://" + searchValue},
+		DisplayName:   searchValue,
+		LoginName:     searchValue,
+		PrincipalType: principalType,
+		Provider:      g.Name,
+	}
+
+	principals = append(principals, p)
+	return principals, nil
+}
+
+func (g *GenOIDCProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	var p v3.Principal
+
+	// parsing id to get the external id and type. Example genericoidc_<user|group>://<user sub | group name>
+	principalScheme, externalID, found := strings.Cut(principalID, "://")
+	if !found {
+		return p, fmt.Errorf("invalid principal id: %s", principalID)
+	}
+	provider, principalType, found := strings.Cut(principalScheme, "_")
+	if !found {
+		return p, fmt.Errorf("invalid principal scheme: %s", principalScheme)
+	}
+
+	if externalID == "" && principalType == "" {
+		return p, fmt.Errorf("invalid id %v", principalID)
+	}
+	if principalType != UserType && principalType != GroupType {
+		return p, fmt.Errorf("invalid principal type: %s", principalType)
+	}
+	if principalType == UserType {
+		p = v3.Principal{
+			ObjectMeta:    metav1.ObjectMeta{Name: provider + "_" + principalType + "://" + externalID},
+			DisplayName:   externalID,
+			LoginName:     externalID,
+			PrincipalType: UserType,
+			Provider:      g.Name,
+		}
+	} else {
+		p = g.groupToPrincipal(externalID)
+	}
+	p = g.toPrincipalFromToken(principalType, p, &token)
+	return p, nil
+}
+
+// TransformToAuthProvider yields information used, typically by the UI, to be able to form URLs used to perform login.
+func (g *GenOIDCProvider) TransformToAuthProvider(authConfig map[string]interface{}) (map[string]interface{}, error) {
+	p := common.TransformToAuthProvider(authConfig)
+	p[publicclient.GenericOIDCProviderFieldRedirectURL] = g.getRedirectURL(authConfig)
+	p[publicclient.GenericOIDCProviderFieldScopes] = authConfig["scope"]
```

Comment:
Do we have any validation to verify that the `openid` scope is included in this list?

Reply:
To be fair, the error the user gets will normally be fairly unambiguous.

But it feels like we could improve this at the activation level for sure.

Reply:
check added